modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/model.scala

```diff
@@ -52,26 +58,37 @@ object QState {
     * If the `Current` doesn't have all actions completed or there are no more
     * pending `Execution`s it returns None.
     */
-  def next(st: QState): Option[(Execution[Action], QState)] = for {
-    exe3done <- uncurrentify(st.current)
-    qd = Queue.cons(exe3done)(st.done)
-    (exe3pending, qp) <- Queue.uncons(st.pending)
-    (actions, c) = currentify(exe3pending)
-  } yield (actions, QState(qp, c, qd, st.status))
+  def next(qs: QState): Option[QState] = cleanup(qs) >>= (prime(_))
+
+  // None: current not empty
+  //       pending queue empty
+  def prime(qs: QState): Option[QState] =
+    if (qs.current.isEmpty)
+      Queue.uncons(qs.pending).map(
+        t => {
+         val (exe3, q) = t
+         QState(q, currentify(exe3), qs.done, qs.status)
+        }
+      )
+    else None
 
+  def cleanup(qs: QState): Option[QState] =
+    uncurrentify(qs.current).map(exe3 =>
+      QState(qs.pending, Current.empty, Queue.cons(exe3)(qs.done), qs.status)
+    )
   /**
     * Transform an *unconsed* pending `Execution` into `Current` in addition to
     * returning the unwrapped `Execution`.
     */
   // TODO: Use same structure for `Current` and `Queue.Execution3`?
-  private def currentify(exe3: Queue.Execution3[Action]): (Execution[Action], Current) = {
+  private def currentify(exe3: Queue.Execution3[Action]): (Current) = {
 
     def vec(exe: Execution[Action]): Vector[Action \/ Result] = exe.map(_.left).toVector
 
     exe3 match {
-      case -\/(-\/((actions, seqid, stepid))) => (actions, Current(vec(actions), Some((seqid, stepid).left)))
-      case -\/(\/-((actions, stepid))) => (actions, Current(vec(actions), Some(stepid.right)))
-      case \/-(actions) => (actions, Current(vec(actions), None))
+      case -\/(-\/((actions, seqid, stepid))) => Current(vec(actions), Some((seqid, stepid).left))
```

Comment:
Each sum type of the `\/` represents if either a `Sequence` or a `Step` or a `Execution` in the current `Step` was completed. It's explained in the documentation for that function but I should put a small comment for each case pattern to make the code easier to follow.

`\/` with more than 2 types was strange for me too because in Haskell you can't do that. One of the nice features of `\/` is that it supports sum types of arbitrary number of elements, however in practice it turns out it's not so great because as you can see, pattern matching is not so straightforward.

The solution for this is to create a custom sum data type. This will be necessary, in any case, when adding more complex Steps so I'm writing this down as an improvement to do later.

In any case this *weird* sum types are internal to this object, they are not meant to be used anywhere else.